components/ai_chat/browser/page_content_fetcher.cc

```diff
@@ -125,16 +127,72 @@ class PageContentFetcher {
                network::SimpleURLLoader::RetryMode::RETRY_ON_NETWORK_CHANGE);
     loader->SetAllowHttpErrorResults(true);
     auto* loader_ptr = loader.get();
-    auto on_response = base::BindOnce(
-        &PageContentFetcher::OnTranscriptFetchResponse,
-        weak_ptr_factory_.GetWeakPtr(), std::move(callback), std::move(loader));
+    bool is_youtube =
+        data->type == ai_chat::mojom::PageContentType::VideoTranscriptYouTube;
+    auto on_response =
+        base::BindOnce(&PageContentFetcher::OnTranscriptFetchResponse,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                       std::move(loader), is_youtube);
     loader_ptr->DownloadToString(url_loader_factory_.get(),
                                  std::move(on_response), 2 * 1024 * 1024);
   }
 
+  void OnYoutubeTranscriptXMLParsed(
+      FetchPageContentCallback callback,
+      base::expected<base::Value, std::string> result) {
+    // Example Youtube transcript XML:
+    //
+    // <?xml version="1.0" encoding="utf-8"?>
+    // <transcript>
+    //   <text start="0" dur="5.1">Dear Fellow Scholars, this is Two Minute
+    //   Papers with Dr. Károly Zsolnai-Fehér.</text> <text start="5.1"
+    //   dur="5.28">ChatGPT has just been supercharged with  browsing support, I
+    //   tried it myself too,  </text> <text start="10.38" dur="7.38">and I
+    //   think this changes everything. Well, almost  everything, as you will
+    //   see in a moment. And this  </text>
+    // </transcript>
+
+    if (!data_decoder::IsXmlElementNamed(result.value(), "transcript")) {
+      return;
+    }
+
+    std::string transcript_text;
+    const base::Value::List* children =
+        data_decoder::GetXmlElementChildren(result.value());
+    if (!children) {
+      return;
+    }
+
+    for (const auto& child : *children) {
+      if (!data_decoder::IsXmlElementNamed(child, "text")) {
+        continue;
+      }
+
+      std::string text;
+      if (!data_decoder::GetXmlElementText(child, &text)) {
+        continue;
+      }
+
+      if (!transcript_text.empty()) {
+        // Add a space as a separator betwen texts.
+        transcript_text += " ";
+      }
+
+      // Remove the two NBSP at the end of each text
+      base::ReplaceFirstSubstringAfterOffset(&text, 0, "  ", "");
+
+      transcript_text += text;
+    }
+
+    // Remove the newlines are scattered throughout.
+    base::RemoveChars(transcript_text, "\n", &transcript_text);
```

Comment:
I wonder where the nbsp and the newlines are coming from - I don't see them in the XML when I do a brief fetch in the browser - but they're always there?

Reply:
I'm also wondering if there's a difference between transcripts that are inferred by YouTube's speech-to-text algorithm vs manual user entry. Perhaps it preserves newlines to keep sentences or paragraphs together? Perhaps it's irrelevant for our AI Chat purposes, or perhaps it would aid the AI's understanding. Just wondering because I don't see them on some of the transcripts.

Reply:
I took a closer look and it appears this may just be an artifact on the particular video I was developing this with, as I wasn't able to find another instance of these new lines and NBSP characters in other videos. I've removed these replacements for that reason.